<a href="https://colab.research.google.com/github/dimLMT/Prediction-of-Product-Sales/blob/main/part2_of_Prediction_of_Product_Sales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prediction of Product Sales
- Author: Mengtian Liu


## Project Overview
- sales prediction for food items sold at various stores. The goal of this is to help the retailer understand the properties of products and outlets that play crucial roles in increasing sales.

## Load and Inspect Data

- Use Pandas to read the sales prediction data set into Google Colab.
  
- Use df.info() and df.head() methods to preview the first rows of your data and a summary of your DataFrame's columns.


In [3]:
# Mount google drive
from google.colab import drive
drive.mount('/content/drive')
#load file
import numpy as np
import pandas as pd

fpath = '/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv'
df = pd.read_csv(fpath)
df.info()
print('\n\n')
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB





Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

## Clean Data
- Explore the data - what do you need to do to clean this data? Clean it!
- Be sure to address each of the following tasks:

1) How many rows and columns?

2) What are the datatypes of each variable?

In [4]:
df.shape

(8523, 12)

In [5]:
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

3) Are there duplicates? If so, drop any duplicates.

4) Identify missing values.

In [6]:
df.duplicated().sum()

0

In [7]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

5) Find and fix any inconsistent categories of data (for example: fix cat, Cat, and cats so that they are consistent).


In [8]:
cat_cols = df.select_dtypes('object').columns
for cols in cat_cols:
  print(df[cols].value_counts(),'\n')

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64 

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64 

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64 

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64 

Mediu

In [9]:
consis_dict = {'LF':'Low Fat', 'reg': 'Regular', 'low fat':'Low Fat'}
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace(consis_dict)
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

6) For any numerical columns, obtain the summary statistics of each (min, max, mean).

In [10]:
df.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

## Start using machine learning to make predictions about future sales based on the data provided.

First, ensure there is no data leakage!

Identify the features (X) and target (y): Assign the "Item_Outlet_Sales" column as your target and the rest of the relevant variables as your features matrix.

Hint: We recommend you drop the "Item_Identifier" feature because it has very high cardinality.

In [11]:
# Set pandas as the default output for sklearn
from sklearn import set_config
set_config(transform_output='pandas')
# other imports
from sklearn.pipeline import make_pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split

# Perform a train test split
y = df['Item_Outlet_Sales']
X = df.drop(columns=['Item_Outlet_Sales','Item_Identifier'])
X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=42)
X_train.head()

Item_Weight Item_Fat_Content  Item_Visibility     Item_Type  Item_MRP  \
4776       16.350          Low Fat         0.029565     Household  256.4646   
7510       15.250          Regular         0.000000   Snack Foods  179.7660   
5828       12.350          Regular         0.158716          Meat  157.2946   
5327        7.975          Low Fat         0.014628  Baking Goods   82.3250   
4810       19.350          Low Fat         0.016645  Frozen Foods  120.9098   

     Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
4776            OUT018                       2009      Medium   
7510            OUT018                       2009      Medium   
5828            OUT049                       1999      Medium   
5327            OUT035                       2004       Small   
4810            OUT045                       2002         NaN   

     Outlet_Location_Type        Outlet_Type  
4776               Tier 3  Supermarket Type2  
7510               Tier 3  Supermarket Type2  
5828               Tier 1  Supermarket Type1  
5327               Tier 2  Supermarket Type1  
4810               Tier 2  Supermarket Type1

Create a preprocessing object to prepare the dataset for Machine Learning.

Make sure your imputation of missing values occurs after the train test split using SimpleImputer.

In [12]:
# PREPROCESSING PIPELINE FOR NUMERIC DATA
num_cols = X_train.select_dtypes("number").columns

# instantiate preprocessors
num_impute_na = SimpleImputer(strategy='median')
num_scaler = StandardScaler()
# Make a numeric preprocessing pipeline
num_pipe = make_pipeline(num_impute_na, num_scaler)
num_pipe

Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median')),
                ('standardscaler', StandardScaler())])

In [13]:
# list of all categorical columns
cat_cols = X_train.select_dtypes("object").columns
cat_cols

Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [14]:
# PREPROCESSING PIPELINE FOR ORDINAL DATA
ord_cols = ['Outlet_Size']
print("Ordinal Columns:", ord_cols)
# Create imputer for ordinal data
ord_impute_na = SimpleImputer(strategy='constant', fill_value='NA')
## Making the OrdinalEncoder
# Specifying order of categories for our Ordinal Columms
OS_order = ['NA','Small','Medium','High']
# Making the list of order lists for OrdinalEncoder
ordinal_orders = [OS_order]
ord_encoder = OrdinalEncoder(categories=ordinal_orders)
# Making a final scaler to scale ordinal data
ord_scaler= StandardScaler()
## Making an ord_pipe
ord_pipe = make_pipeline(ord_impute_na, ord_encoder, ord_scaler)
ord_pipe

Ordinal Columns: ['Outlet_Size']


Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='NA', strategy='constant')),
                ('ordinalencoder',
                 OrdinalEncoder(categories=[['NA', 'Small', 'Medium',
                                             'High']])),
                ('standardscaler', StandardScaler())])

In [15]:
# PREPROCESSING PIPELINE FOR ONE-HOT-ENCODED DATA
nom_cols = X_train.select_dtypes('object').drop(columns=ord_cols).columns
print("Nominal Columns:", nom_cols)
# Instantiate the individual preprocessors
nom_impute_na = SimpleImputer(strategy='constant', fill_value = "NA")
nom_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
# Make pipeline with imputer and encoder
nom_pipe = make_pipeline(nom_impute_na, nom_encoder)
nom_pipe


Nominal Columns: Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')


Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='NA', strategy='constant')),
                ('onehotencoder',
                 OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

In [16]:
# Making a numeric tuple for ColumnTransformer
num_tuple = ('numeric', num_pipe, num_cols)
print(num_tuple)
# Making an ordinal_tuple for ColumnTransformer
ord_tuple = ('ordinal', ord_pipe, ord_cols)
print(ord_tuple)
# Making a nominal_tuple for ColumnTransformer
nom_tuple = ('nominal', nom_pipe, nom_cols)
print(nom_tuple)

('numeric', Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median')),
                ('standardscaler', StandardScaler())]), Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object'))
('ordinal', Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='NA', strategy='constant')),
                ('ordinalencoder',
                 OrdinalEncoder(categories=[['NA', 'Small', 'Medium',
                                             'High']])),
                ('standardscaler', StandardScaler())]), ['Outlet_Size'])
('nominal', Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='NA', strategy='constant')),
                ('onehotencoder',
                 OneHotEncoder(handle_unknown='ignore', sparse_output=False))]), Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))


In [17]:
# Instantiate with verbose_feature_names_out=False
col_transformer = ColumnTransformer([num_tuple, ord_tuple, nom_tuple],
                                    verbose_feature_names_out=False)
# Fit on training data
col_transformer.fit(X_train)

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('ordinal',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='NA',
                                                                strategy='constant')),
                                                 ('ordinale...
                                                  StandardScaler())]),
                                 ['Outlet_Size']),
                                ('nominal',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='NA',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                  verbose_feature_names_out=False)

In [18]:
# Transform the training data
X_train_processed = col_transformer.transform(X_train)
# Transform the testing data
X_test_processed = col_transformer.transform(X_test)
# View the processed training data
X_train_processed.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
4776     0.827485        -0.712775  1.828109                   1.327849   
7510     0.566644        -1.291052  0.603369                   1.327849   
5828    -0.121028         1.813319  0.244541                   0.136187   
5327    -1.158464        -1.004931 -0.952591                   0.732018   
4810     1.538870        -0.965484 -0.336460                   0.493686   

      Outlet_Size  Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
4776     0.748125                       1.0                       0.0   
7510     0.748125                       0.0                       1.0   
5828     0.748125                       0.0                       1.0   
5327    -0.264370                       1.0                       0.0   
4810    -1.276865                       1.0                       0.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  ...  \
4776                     0.0               0.0                  0.0  ...   
7510                     0.0               0.0                  0.0  ...   
5828                     0.0               0.0                  0.0  ...   
5327                     1.0               0.0                  0.0  ...   
4810                     0.0               0.0                  0.0  ...   

      Outlet_Identifier_OUT045  Outlet_Identifier_OUT046  \
4776                       0.0                       0.0   
7510                       0.0                       0.0   
5828                       0.0                       0.0   
5327                       0.0                       0.0   
4810                       1.0                       0.0   

      Outlet_Identifier_OUT049  Outlet_Location_Type_Tier 1  \
4776                       0.0                          0.0   
7510                       0.0                          0.0   
5828                       1.0                          1.0   
5327                       0.0                          0.0   
4810                       0.0                          0.0   

      Outlet_Location_Type_Tier 2  Outlet_Location_Type_Tier 3  \
4776                          0.0                          1.0   
7510                          0.0                          1.0   
5828                          0.0                          0.0   
5327                          1.0                          0.0   
4810                          1.0                          0.0   

      Outlet_Type_Grocery Store  Outlet_Type_Supermarket Type1  \
4776                        0.0                            0.0   
7510                        0.0                            0.0   
5828                        0.0                            1.0   
5327                        0.0                            1.0   
4810                        0.0                            1.0   

      Outlet_Type_Supermarket Type2  Outlet_Type_Supermarket Type3  
4776                            1.0                            0.0  
7510                            1.0                            0.0  
5828                            0.0                            0.0  
5327                            0.0                            0.0  
4810                            0.0                            0.0  

[5 rows x 40 columns]

### Build a linear regression model.
- Use the custom evaluation function to get the metrics for your model (on training and test data).
- Compare the training vs. test R-squared values and answer the question: to what extent is this model overfit/underfit?

In [19]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor

# the custom evaluation function to get the metrics for your model
def regression_metrics(y_true, y_pred, label='', verbose = True, output_dict=False):
  # Get metrics
  mae = mean_absolute_error(y_true, y_pred)
  mse = mean_squared_error(y_true, y_pred)
  rmse = mean_squared_error(y_true, y_pred, squared=False)
  r_squared = r2_score(y_true, y_pred)
  if verbose == True:
    # Print Result with Label and Header
    header = "-"*60
    print(header, f"Regression Metrics: {label}", header, sep='\n')
    print(f"- MAE = {mae:,.3f}")
    print(f"- MSE = {mse:,.3f}")
    print(f"- RMSE = {rmse:,.3f}")
    print(f"- R^2 = {r_squared:,.3f}")
  if output_dict == True:
      metrics = {'Label':label, 'MAE':mae,
                 'MSE':mse, 'RMSE':rmse, 'R^2':r_squared}
      return metrics

def evaluate_regression(reg, X_train, y_train, X_test, y_test, verbose = True,
                        output_frame=False):
  # Get predictions for training data
  y_train_pred = reg.predict(X_train)

  # Call the helper function to obtain regression metrics for training data
  results_train = regression_metrics(y_train, y_train_pred, verbose = verbose,
                                     output_dict=output_frame,
                                     label='Training Data')
  # Get predictions for test data
  y_test_pred = reg.predict(X_test)
  # Call the helper function to obtain regression metrics for test data
  results_test = regression_metrics(y_test, y_test_pred, verbose = verbose,
                                  output_dict=output_frame,
                                    label='Test Data' )

  # Store results in a dataframe if ouput_frame is True
  if output_frame:
    results_df = pd.DataFrame([results_train,results_test])
    # Set the label as the index
    results_df = results_df.set_index('Label')
    # Set index.name to none to get a cleaner looking result
    results_df.index.name=None
    # Return the dataframe
    return results_df.round(3)

In [20]:
# Instantiate a linear regression model
linreg = LinearRegression()
# Combine the preprocessing ColumnTransformer and the linear regression model in a Pipeline
linreg_pipe = make_pipeline(col_transformer, linreg)
linreg_pipe

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('numeric',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('standardscaler',
                                                                   StandardScaler())]),
                                                  Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                                 ('ordinal',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value=...
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value='NA',
                                                                                 strategy='constant')),
                                                                  ('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                                   verbose_feature_names_out=False)),
                ('linearregression', LinearRegression())])

In [21]:
# Fit the model pipeline on the training data
linreg_pipe.fit(X_train, y_train)
# Obtain Model Evulation using custom function
evaluate_regression(linreg_pipe, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 847.120
- MSE = 1,297,558.548
- RMSE = 1,139.104
- R^2 = 0.562
------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 804.048
- MSE = 1,194,230.406
- RMSE = 1,092.808
- R^2 = 0.567


$\color{blue}{this\ model\ is\ underfit\ due\ to\ low\ R^2\ values }$
### Build a default Random Forest model.
- Use the custom evaluation function to get the metrics for your model (on training and test data).
- Compare the training vs. test R-squared values and answer the question: to what extent is this model overfit/underfit?
- Compare this model's performance to the linear regression model: which model has the best test scores?

In [22]:
# Instantiate default random forest model
rfreg = RandomForestRegressor(random_state = 42)
rfreg_pipe = make_pipeline(col_transformer, rfreg)
rfreg_pipe

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('numeric',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('standardscaler',
                                                                   StandardScaler())]),
                                                  Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                                 ('ordinal',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value=...
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value='NA',
                                                                                 strategy='constant')),
                                                                  ('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                                   verbose_feature_names_out=False)),
                ('randomforestregressor',
                 RandomForestRegressor(random_state=42))])

In [23]:
# Fit the model pipeline and Obtain Model Evulation
# Fit the model pipeline on the training data only
rfreg_pipe.fit(X_train, y_train)
evaluate_regression(rfreg_pipe, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 296.120
- MSE = 182,642.045
- RMSE = 427.366
- R^2 = 0.938
------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 767.081
- MSE = 1,216,212.220
- RMSE = 1,102.820
- R^2 = 0.559


$\color{blue}{this\ model\ is\ overfit\ due\ to\ the\ high\ R^2\ value\ for\ training\ data\ only\ }$
- the linear regression model is slightly better here, with the R^2 of 0.567
###  Use GridSearchCV to tune at least two hyperparameters for a Random Forest model.

After determining the best parameters from your GridSearch, fit and evaluate a final best model on the entire training set (no folds).

Compare your tuned model to your default Random Forest: did the performance improve?

In [30]:
# Define param grid with options to try
params = {'randomforestregressor__n_estimators':[50,100,150],
          'randomforestregressor__max_features':['sqrt',None],
          'randomforestregressor__max_depth': [None,4,5,6,10],
          'randomforestregressor__min_samples_leaf':[1,2]}

In [31]:
# Instantiate the gridsearch
gridsearch = GridSearchCV(rfreg_pipe, params, n_jobs=-1, cv = 3, verbose=1)
# Fit the gridsearch on training data
gridsearch.fit(X_train, y_train)
gridsearch.best_params_

Fitting 3 folds for each of 60 candidates, totalling 180 fits


{'randomforestregressor__max_depth': 6,
 'randomforestregressor__max_features': None,
 'randomforestregressor__min_samples_leaf': 1,
 'randomforestregressor__n_estimators': 100}

In [32]:
## Evaluate the best model
best_model = gridsearch.best_estimator_
evaluate_regression(best_model, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 742.153
- MSE = 1,111,128.518
- RMSE = 1,054.101
- R^2 = 0.625
------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 727.510
- MSE = 1,099,976.656
- RMSE = 1,048.798
- R^2 = 0.601


Surprisingly, the performance did not improve!

- You need to determine which model to implement. Overall, which model do you recommend?

```Random Forest Regressor```

- Justify your recommendation. Interpret your model's performance based on R-squared in a way that your non-technical stakeholder can understand.

```0.601 is still slighlty larger than the other two model```

- Select another regression metric (RMSE/MAE/MSE) to express the performance of your model to your stakeholder. Include why you selected this metric to explain to your stakeholder.

```RMSE, a larger RMSE means more error/deviation from the true data. The prediction is that Item_Outlet_Sales will be within the range of Item_Outlet_Sales - RMSE to Item_Outlet_Sales + RMSE, so a smaller RMSE gives better prediction.```

- Compare the training vs. test scores and answer the question: to what extent is this model overfit/underfit?

```The tuned random forest model remains underfit, as ~0.6 is not a ideal R^2 value. It is reccomened that additional models be explored.```